File: db_builder/dag_steamapi.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// pulls from steam store page based on the unique game id
func steamApiPull(appID int) (
	genre, description, website, headerImage, background, screenshot, steamURL, pricing, achievements string,
) {
	str := strconv.Itoa(appID)
	url := fmt.Sprintf("https://store.steampowered.com/api/appdetails?appids=%s", str)

	maxRetries := 3
	for attempt := 0; attempt < maxRetries; attempt++ {
		resp, err := http.Get(url)
		if err != nil {
			log.Printf("Failed to fetch data for app %d: %v", appID, err)
			return
		}
		defer resp.Body.Close()

		// If rate limited, wait and retry
		if resp.StatusCode == 429 {
			retryAfter := resp.Header.Get("Retry-After")
			wait := 30 * time.Second // Reduced default wait time

			if retryAfter != "" {
				if seconds, err := strconv.Atoi(retryAfter); err == nil {
					wait = time.Duration(seconds) * time.Second
				}
			}

			log.Printf("Rate limited for app %d. Waiting %v before retry %d/%d",
				appID, wait, attempt+1, maxRetries)
			time.Sleep(wait)
			continue
		}

		// Check other bad statuses
		if resp.StatusCode != http.StatusOK {
			log.Printf("Bad status for app %d: %s", appID, resp.Status)
			return
		}

		// Read the response body
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			log.Printf("Failed to read response for app %d: %v", appID, err)
			return
		}

		// Check if response starts with HTML (error page)
		if strings.HasPrefix(string(body), "<") {
			log.Printf("Got HTML response for app %d, likely rate limited or invalid ID", appID)
			time.Sleep(5 * time.Second) // Brief pause before retry
			continue
		}

		var result map[string]struct {
			Success bool `json:"success"`
			Data    struct {
				ShortDesc   string `json:"short_description"`
				Website     string `json:"website"`
				HeaderImage string `json:"header_image"`
				Background  string `json:"background"`
				Genres      []struct {
					Description string `json:"description"`
				} `json:"genres"`
				Screenshots []struct {
					PathFull string `json:"path_full"`
				} `json:"screenshots"`
				PriceOverview struct {
					FinalFormatted string `json:"final_formatted"`
				} `json:"price_overview"`
				Achievements struct {
					Total int `json:"total"`
				} `json:"achievements"`
			} `json:"data"`
		}

		if err := json.Unmarshal(body, &result); err != nil {
			log.Printf("Failed to parse JSON for app %d: %v", appID, err)
			return
		}

		// Check if the API call was successful
		appData, exists := result[str]
		if !exists || !appData.Success {
			log.Printf("API returned success=false for app %d", appID)
			return
		}

		info := appData.Data

		// Join genres
		var genreList []string
		for _, g := range info.Genres {
			genreList = append(genreList, g.Description)
		}
		genre = strings.Join(genreList, ", ")

		// First screenshot if exists
		if len(info.Screenshots) > 0 {
			screenshot = info.Screenshots[0].PathFull
		}

		// Steam store URL
		steamURL = fmt.Sprintf("https://store.steampowered.com/app/%s", str)

		// Achievements to string
		if info.Achievements.Total > 0 {
			achievements = fmt.Sprintf("%d achievements", info.Achievements.Total)
		} else {
			achievements = "0 achievements"
		}

		return genre, info.ShortDesc, info.Website, info.HeaderImage, info.Background,
			screenshot, steamURL, info.PriceOverview.FinalFormatted, achievements
	}

	return // Return empty if all retries failed
}

File: database_builder/dag_steamapi.py
import requests
import json
import time
import logging

def steam_api_pull(app_id):
    url = f"https://store.steampowered.com/api/appdetails?appids={app_id}"
    
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = requests.get(url)
            
            if response.status_code == 429:
                retry_after = response.headers.get('Retry-After')
                wait_time = 30
                
                if retry_after:
                    try:
                        wait_time = int(retry_after)
                    except ValueError:
                        pass
                
                logging.warning(f"Rate limited for app {app_id}. Waiting {wait_time}s before retry {attempt+1}/{max_retries}")
                time.sleep(wait_time)
                continue
            
            if response.status_code != 200:
                logging.error(f"Bad status for app {app_id}: {response.status_code}")
                return "", "", "", "", "", "", "", "", ""
            
            if response.text.startswith("<"):
                logging.warning(f"Got HTML response for app {app_id}, likely rate limited or invalid ID")
                time.sleep(5)
                continue
            
            try:
                data = response.json()
            except json.JSONDecodeError:
                logging.error(f"Failed to parse JSON for app {app_id}")
                return "", "", "", "", "", "", "", "", ""
            
            app_data = data.get(str(app_id))
            if not app_data or not app_data.get('success'):
                logging.warning(f"API returned success=false for app {app_id}")
                return "", "", "", "", "", "", "", "", ""
            
            info = app_data.get('data', {})
            
            genre_list = [g['description'] for g in info.get('genres', [])]
            genre = ", ".join(genre_list)
            
            description = info.get('short_description', '')
            website = info.get('website', '')
            header_image = info.get('header_image', '')
            background = info.get('background', '')
            
            screenshot = ""
            screenshots = info.get('screenshots', [])
            if screenshots:
                screenshot = screenshots[0].get('path_full', '')
            
            steam_url = f"https://store.steampowered.com/app/{app_id}"
            
            price_overview = info.get('price_overview', {})
            pricing = price_overview.get('final_formatted', '')
            
            achievements_data = info.get('achievements', {})
            total_achievements = achievements_data.get('total', 0)
            achievements = f"{total_achievements} achievements" if total_achievements > 0 else "0 achievements"
            
            return genre, description, website, header_image, background, screenshot, steam_url, pricing, achievements
            
        except requests.RequestException as e:
            logging.error(f"Request failed for app {app_id}: {e}")
            if attempt < max_retries - 1:
                time.sleep(5)
                continue
            return "", "", "", "", "", "", "", "", ""
    
    return "", "", "", "", "", "", "", "", ""